runtime/v2: add note about orphan process for runc-shim (#10002)
containerd/containerd: core/runtime/v2/README.md

The only difference between grpc and ttrpc is the wire protocol.
ttrpc removes the http stack in order to save memory and binary size to keep shims small.
It is recommended to use ttrpc in your shim but grpc support is currently an experimental feature.

#### containerd-shim-runc-v2 as sub-reaper
The shim process takes responsibility as a sub-reaper to cleanup exited containers or setns(2) processes.

> When container is running in new PID namespace, the container should cleanup orphaned processes before it exits.

Reviewer: ```suggestion
When a container runs in a new PID namespace, the container should clean up orphaned processes before it exits.
```

> If container uses the same PID namespace with shim process, its descendant processes will be reparented to shim process. The shim process will reap them when they exit.

Reviewer: ```suggestion
If the container uses the same PID namespace as the shim process, its descendant processes will be reparented to the shim process. The shim process will then reap them upon their exit.
```

However, [\[PATCH\] exit: fix the setns() && PR_SET_CHILD_SUBREAPER interaction](https://lore.kernel.org/all/20170130181735.GA11285@redhat.com/#r) prevents any cross-namespace reparenting in kernel. Assume that container is in X-namespace and P in root-namespace setns into X-namespace. P forks child C. The child C forks a grandchild G and exits. The G will be reparented to X instead of P's reaper.

> If the PID namespace is different from shim process, the container init process should cleanup any orphaned reparented processes created by setns process (exec operation).

Reviewer: ```suggestion
If the PID namespace differs from the shim process, the container's init process should ensure the cleanup of any orphaned, reparented processes created by the setns operation (e.g., during an exec).
```

The only difference between grpc and ttrpc is the wire protocol.
ttrpc removes the http stack in order to save memory and binary size to keep shims small.
It is recommended to use ttrpc in your shim but grpc support is currently an experimental feature.

> #### runc-shim reap orphan process created by exec-init

Reviewer 2: Suggestion:

```
#### containerd-shim-runc-v2 as sub-reaper

The shim process takes responsibility as a sub-reaper to cleanup exited containers or setns(2) processes.

When container is running in new PID namespace, the container should cleanup orphaned processes before it exits.
If container uses the same PID namespace with shim process, its descendant processes will be reparented to shim process. The shim process will reap them when they exit.

However, [[PATCH] exit: fix the setns() && PR_SET_CHILD_SUBREAPER interaction](https://lore.kernel.org/all/20170130181735.GA11285@redhat.com/#r) prevents any cross-namespace reparenting.

If the PID namespace is different from shim process, the container init process should cleanup any orphaned reparented processes created by setns process (exec operation).
```

cc @AkihiroSuda @dmcgowan

> After this commit https://github.com/torvalds/linux/commit/c6c70f4455d1eda91065e93cc4f7eddf4499b105 (merged in kernel 4.11)

Reviewer 2: As far as I know, the shim is not designed to reap the orphan process created by exec-init.
However, the kernel has bug and it's like bug used as a feature for someone.
I'm not sure it's good place to highlight tini or pod-level pid namespace here.

Reviewer 2: Yeah, we can highlight shim doesn't have responsibility to reap any orphan processes created by exec-init. And it's just for runc-shim.

Author: sometimes user use different kernel  have different behaviors ,they don't know why. I think we can let user know why

Author: > Yeah, we can highlight shim doesn't have responsibility to reap any orphan processes created by exec-init. And it's just for runc-shim.
runc-shim has process to reap orphan process but  doesn't work after kernel 4.11.
https://github.com/containerd/containerd/blob/b40a912e7b42c47bf06068508fe948840a80730b/pkg/shim/shim_unix.go#L85

Reviewer 2: The reaper is for init container or exec-init, not for the orphan process.

Author: > The reaper is for init container or exec-init, not for the orphan process.

you are right, shim will  reap orphan process created by exec-init before kernel 4.11.
